### ISSUE TITLE
useSWRMutation is not working with React Native New Architecture.

### ISSUE THREAD

**Issue by ShineKidd** [vercel/swr#2986]
# Bug report

## Repro Steps / Code Example
1. npx create-expo-app@latest

configure app.json 

``` json 
 "plugins": [
      "expo-router",
      [
        "expo-build-properties",
        {
          "ios": {
            "newArchEnabled": true
          },
          "android": {
            "newArchEnabled": true
          }
        }
      ]
    ],
```

3. home.tsx code example

```tsx
import useSWRMutation from 'swr/mutation'

function Home () {
  const {trigger, data, error} = useSWRMutation('https://github.com', (key) => fetch(key).then(res => res.text()))

  console.log({data, error}) // always print {"data": undefined, "error": undefined}

  function submit () {
    // print error: "(NOBRIDGE) ERROR  [TypeError: Cannot read property 'add' of undefined]"
    trigger().then(console.log).catch(console.error)
  }

  return  (
     <View>
       <Button title='mutation' onPress={submit}></Button>
     </View>
     )
}
```

**Comment by idrakimuhamad** [comment]
I've facing the same issue whenever using the mutation. Probably related to the usage of transition.

<img width="795" alt="image" src="https://github.com/user-attachments/assets/062a1572-a67d-4373-8262-075ef59e9302">

**Comment by raajnadar** [comment]
This issue requires attention we are unable to work with Expo Go, new architecture is enabled on Expo Go.

**Comment by raajnadar** [comment]
Found this ussue ticket - https://github.com/facebook/react-native/issues/47730#issuecomment-2495544910

**Comment by peterhpchen** [comment]
I encountered this issue as well. Before resolving it on Expo Go, my workaround was to use a [development build](https://docs.expo.dev/develop/development-builds/create-a-build/) to continue the development process.

**Comment by raajnadar** [comment]
So you are saying it works with the same code used with a dev build? @peterhpchen

**Comment by peterhpchen** [comment]
@raajnadar . Yes, I didn't change any code from Expo Go to Development Build.

The below is my steps:

1. Build: `eas build --profile development --platform ios --local`
2. Run: `npx expo start --dev-client`

My `eas.json`:
```json
{
  "build": {
    "development": {
      "developmentClient": true,
      "distribution": "internal",
      "android": {
        "gradleCommand": ":app:assembleDebug"
      },
      "ios": {
        "buildConfiguration": "Debug"
      }
    },
  },
}
```

**Comment by raajnadar** [comment]
Did you try running it with a preview build, an apk or a test flight build? Because that did not work for me, let me check with the dev client build.

**Comment by peterhpchen** [comment]
@raajnadar, sorry for the confusion. After some configuration changes, I realized I had disabled the New Architecture in the development build. When I enable the New Architecture (`"newArchEnabled": true`), I encounter the same `Cannot read property 'add' of undefined` error.

ref: https://docs.expo.dev/guides/new-architecture/#disable-the-new-architecture-in-an-existing-project

**Comment by raajnadar** [comment]
So we can surely say the issue is with the new architecture and also it is in core react native, there is no problem in SWR library.

Please share any feedback you have in the React native thread, I created a thread to follow up and escalate the issue.

https://github.com/facebook/react-native/issues/47947

**Comment by geryit** [comment]
Seems like it works in new arch without `data` and `isMutating`

This works: `const { trigger } = useSWRMutation(...` but this doesnt `const { trigger, data } = useSWRMutation...`

So my solution was using `useSWRMutation` like this:

```tsx
const [data, setData] = useState(null);
const { trigger } = useSWRMutation('/api/user', updateUser);

useEffect(()=>{
 trigger().then(res=>setData(res))
}, [])

```

**Comment by raajnadar** [comment]
The whole point of SWR was to avoid the usage of useEffect by devs for data fetching and to prevent edge cases & memory leaks in the application.

**Comment by goksel-p** [comment]
> The whole point of SWR was to avoid the usage of useEffect by devs for data fetching and to prevent edge cases & memory leaks in the application.

yes, this is a quick solution for the devs who use the new arch and dont want to migrate all triggers

**Comment by raajnadar** [comment]
You can use this patch this works for me

```
diff --git a/node_modules/swr/dist/mutation/index.js b/node_modules/swr/dist/mutation/index.js
index 0e34690..8607e14 100644
--- a/node_modules/swr/dist/mutation/index.js
+++ b/node_modules/swr/dist/mutation/index.js
@@ -542,9 +542,9 @@ _internal.OBJECT.defineProperty(_internal.SWRConfig, 'defaultValue', {
  * ```
  */ const useSWR = _internal.withArgs(useSWRHandler);
 
-const startTransition = _internal.IS_REACT_LEGACY ? (cb)=>{
-    cb();
-} : ReactExports__default.default.startTransition;
+// const startTransition = _internal.IS_REACT_LEGACY ? (cb)=>{
+//     cb();
+// } : ReactExports__default.default.startTransition;
 /**
  * An implementation of state with dependency-tracking.
  */ const useStateWithDeps = (state)=>{
@@ -610,6 +610,7 @@ const startTransition = _internal.IS_REACT_LEGACY ? (cb)=>{
 };
 
 const mutation = ()=>(key, fetcher, config = {})=>{
+        const [, startTransition] = ReactExports.useTransition()
         const { mutate } = _internal.useSWRConfig();
         const keyRef = ReactExports.useRef(key);
         const fetcherRef = ReactExports.useRef(fetcher);

```

**Comment by bjutkoski** [comment]
any news? same issue here

**Comment by geriko2000** [comment]
Why this bug still not fixed?

**Comment by koba04** [comment]
I don't understand the cause of the issue and this implementation doesn't have any problem on other environments, why doesn't this work on the new architecture?

**Comment by jcruzv-prog** [comment]
The patch also worked for me, just added in node_modules/swr/dist/mutation/index.js

//find this line
const mutation = ()=>(key, fetcher, config = {})=>{
//and added below
   const [, startTransition] = React.useTransition();

**Comment by sawa-ko** [comment]
I have the same error and this fix the error: https://github.com/vercel/swr/issues/2986#issuecomment-2561904700

Thank you and I hope the swr team can do something about it.

**Comment by den-ii** [comment]
> The patch also worked for me, just added in node_modules/swr/dist/mutation/index.js
> 
> //find this line const mutation = ()=>(key, fetcher, config = {})=>{ //and added below const [, startTransition] = React.useTransition();

patch doesn't work well. I'll switch to react-query

**Comment by lovetingyuan** [comment]
+1
isMutating is always true and the data is always undefined after calling trigger(),
I use expo 52 with newArchEnabled

**Comment by caioedut** [comment]
Same here.

**Comment by geriko2000** [comment]
I think, developers won't fix this. I fixed swr like in comment above and vendor it in my project

**Comment by gabriel-peter** [comment]
This issue is brutal! Please fix @team

**Comment by kuriel-dev** [comment]
> Seems like it works in new arch without `data` and `isMutating`
> 
> This works: `const { trigger } = useSWRMutation(...` but this doesnt `const { trigger, data } = useSWRMutation...`
> 
> So my solution was using `useSWRMutation` like this:
> 
> const [data, setData] = useState(null);
> const { trigger } = useSWRMutation('/api/user', updateUser);
> 
> useEffect(()=>{
>  trigger().then(res=>setData(res))
> }, [])

i tested this "solution", is not working.

**Comment by jcruzv-prog** [comment]
"The error 'Cannot read property 'add' of undefined' no longer appears in my project after updating SWR to the latest version (v2.3.4)."

**Comment by caioedut** [comment]
> "The error 'Cannot read property 'add' of undefined' no longer appears in my project after updating SWR to the latest version (v2.3.4)."

for me too

**Comment by QZIG** [comment]
update to v2.3.4 did not work for me...

> You can use this patch this works for me
> 
> ```
> diff --git a/node_modules/swr/dist/mutation/index.js b/node_modules/swr/dist/mutation/index.js
> index 0e34690..8607e14 100644
> --- a/node_modules/swr/dist/mutation/index.js
> +++ b/node_modules/swr/dist/mutation/index.js
> @@ -542,9 +542,9 @@ _internal.OBJECT.defineProperty(_internal.SWRConfig, 'defaultValue', {
>   * ```
>   */ const useSWR = _internal.withArgs(useSWRHandler);
>  
> -const startTransition = _internal.IS_REACT_LEGACY ? (cb)=>{
> -    cb();
> -} : ReactExports__default.default.startTransition;
> +// const startTransition = _internal.IS_REACT_LEGACY ? (cb)=>{
> +//     cb();
> +// } : ReactExports__default.default.startTransition;
>  /**
>   * An implementation of state with dependency-tracking.
>   */ const useStateWithDeps = (state)=>{
> @@ -610,6 +610,7 @@ const startTransition = _internal.IS_REACT_LEGACY ? (cb)=>{
>  };
>  
>  const mutation = ()=>(key, fetcher, config = {})=>{
> +        const [, startTransition] = ReactExports.useTransition()
>          const { mutate } = _internal.useSWRConfig();
>          const keyRef = ReactExports.useRef(key);
>          const fetcherRef = ReactExports.useRef(fetcher);
> ```

btw the patch workaround solves the pb

**Comment by YashTotale** [comment]
Same error for me for `swr@2.3.4`

**Comment by lovetingyuan** [comment]
@shuding @huozhi  please fix this 🙏

**Comment by raajnadar** [comment]
Try with the latest Expo SDK, it works on Snack Expo with SDK 53
